comit/src/expiries.rs

```diff
@@ -0,0 +1,488 @@
+    BetaRedeemTransactionSeen, // Implies either one confirmation or unconfirmed transaction seen.
+    RedeemAlphaTransactionSent,
+    Done, // Implies alpha transaction confirmed configured number of times.
+}
+
+/// Possible next action for Alice.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum BobAction {
+    // Happy path actions for Bob.
+    WaitForAlphaFunded,
+    FundBeta,
+    WaitForBetaFundTransactionToBeConfirmed,
+    WaitForBetaRedeemTransactionToBeSeen,
+    RedeemAlpha,
+    WaitForAlphaRedeemTransactionToBeConfirmed,
+    NoFurtherAction, // Implies swap is done from Bob's perspective.
+    // Cancel path.
+    Abort,         // Implies HTLC not funded, abort but take no further action.
+    WaitForRefund, // Implies alpha ledger HTLC funded but expiry not yet elapsed.
+    Refund,        // Implies HTLC funded and expiry time elapsed.
+}
+
+impl BobState {
+    /// Gets the next state and the action required to transition to that state.
+    fn next(&self) -> (BobAction, BobState) {
+        use self::{BobAction::*, BobState::*};
+
+        match self {
+            Created => (WaitForAlphaFunded, AlphaFunded),
+            AlphaFunded => (FundBeta, BetaFundTransactionSent),
+            BetaFundTransactionSent => (WaitForBetaFundTransactionToBeConfirmed, BetaFunded),
+            BetaFunded => (
+                WaitForBetaRedeemTransactionToBeSeen,
+                BetaRedeemTransactionSeen,
+            ),
+            BetaRedeemTransactionSeen => (RedeemAlpha, RedeemAlphaTransactionSent),
+            RedeemAlphaTransactionSent => (WaitForAlphaRedeemTransactionToBeConfirmed, Done),
+            Done => (NoFurtherAction, Done),
+        }
+    }
+
+    /// The minimum time we need to allow to transition to the next state.
+    fn transition_period(&self) -> Duration {
+        use self::BobAction::*;
+
+        let (next_action, _next_state) = self.next();
+        if next_action == NoFurtherAction {
+            return Duration::zero();
+        }
+
+        unimplemented!()
+    }
+
+    /// True if Bob has funded the beta HTLC.
+    fn includes_fund(&self) -> bool {
+        match self {
+            BobState::Created => false,
+            BobState::AlphaFunded => false,
+            // Be conservative because whether or not to try and refund depends on this value.
+            BobState::BetaFundTransactionSent => true,
+            BobState::BetaFunded => true,
+            BobState::BetaRedeemTransactionSeen => true,
+            BobState::RedeemAlphaTransactionSent => true,
+            BobState::Done => true,
+        }
+    }
+}
+
+/// To<Role>State returns either our own state or our best guess at
+/// the counterparties state. Returns the latest possible state that
+/// we can guarantee assuming the counterparty has the same blockchain
+/// knowledge as we do.
+
+pub trait AsAliceState {
+    fn as_alice_state(&self) -> AliceState;
+}
```

Comment:
`AsXYZ` usually returns a reference by convention. If you want that, you could also implement `AsRef` :)

Otherwise your doc comment says the right thing: It should be `ToXYZ`.

Reply:
I researched this last time we discussed it, the conclusion I came to was that `AsXYZ` is for 'cheap' things and 'ToXYZ' is for 'expensive' things. With cheap and expensive being defined the same as for `Copy` vs `Clone`. Whether or not it takes/returns a reference is explicit in the method. 

So in this case, the `AliceState` struct implements `Copy` so I elected for `AsAliceState`.